torch/csrc/jit/tensorexpr/block_codegen.cpp

```diff
@@ -0,0 +1,319 @@
+  auto blck_sz = block_analysis_->block_size();
+  std::unordered_set<const Buf*> bufs(buf_reads.begin(), buf_reads.end());
+  bufs.insert(buf_writes.begin(), buf_writes.end());
+
+  if (loop_options.is_gpu_block_index()) {
+    emitIndent();
+    PrintTensorInfo(bufs);
+    PrintDistribution(bufs);
+    PrintBufferInfo(buf_reads);
+    PrintArguments(bufs);
+
+    emitIndent();
+    os() << "compute {" << std::endl;
+
+    PrintReshapeInfo(bufs);
+
+    emitIndent();
+    PrintLoop(bufs, true);
+    v->body()->accept(this);
+
+    os() << std::endl;
+    emitIndent();
+    PrintReshapeInfo(buf_writes, true); // print reverse reshape
+    os() << "}";
+    os() << std::endl;
+  } else if (loop_options.is_gpu_thread_index()) {
+    PrintDMAs(buf_reads);
+    PrintLoop(buf_reads, false);
+    v->body()->accept(this);
+    os() << std::endl;
+    PrintAdjustBuffers(buf_reads);
+
+  } else {
+    IRPrinter::visit(v);
+  }
+}
+
+void BlockPrinter::PrintTensorInfo(std::unordered_set<const Buf*> bufs) {
+  os() << "tensors {";
+  for (const auto& buf : bufs) {
+    os() << std::endl;
+    emitIndent();
+    emitIndent();
+    auto num_dims = block_analysis_->getMultiDimBuf(buf)->dims().size();
+    os() << block_analysis_->getInputName(buf) << " = ";
+    os() << "{";
+    for (auto d = 0; d < num_dims; d++) {
+      os() << "{" << dim_names[d] << "};";
+    }
+    os() << " elem : " << blockDtypeCppString(buf->dtype());
+    os() << "}";
+  }
+
+  for (const auto& buf : bufs) {
+    os() << std::endl;
+    emitIndent();
+    emitIndent();
+    auto num_dims = block_analysis_->getMultiDimBuf(buf)->dims().size();
+    os() << block_analysis_->getFlatInputName(buf) << " = ";
+    os() << "{";
+    os() << "{" << flat_dim_names[num_dims - 1] << "};";
+    os() << " elem : " << blockDtypeCppString(buf->dtype());
+    os() << "}"
+         << " // flattened tensor";
+  }
+  os() << std::endl;
+  emitIndent();
+  os() << "}" << std::endl << std::endl;
+}
+
+void BlockPrinter::PrintArguments(std::unordered_set<const Buf*> bufs) {
+  for (const auto& buf : bufs) {
+    auto multidimbuf = block_analysis_->getMultiDimBuf(buf);
+    auto num_dims = multidimbuf->dims().size();
+
+    // The dims for the multi-dim tensors
+    for (auto d = 0; d < num_dims; d++) {
+      auto dim_val = dynamic_cast<const IntImm*>(multidimbuf->dim(d));
+      this->dim_values_map.emplace(this->dim_names[d], dim_val->value());
+    }
+
+    // The dimensions for the flattened tensors
+    auto val = dynamic_cast<const IntImm*>(buf->dim(0));
+    if (block_analysis_->is_buf_store_target(buf)) {
+      this->dim_values_map.emplace(
+          this->flat_dim_names[num_dims - 1], val->value());
+    }
+  }
+
+  emitIndent();
+  os() << "arguments {" << std::endl;
+
+  for (auto const& arg : this->dim_values_map) {
+    emitIndent();
+    emitIndent();
+    os() << "var " << arg.first << " = " << arg.second << std::endl;
+  }
+
+  emitIndent();
+  emitIndent();
```

Comment:
Why do we need a double indent? If it's needed for correctness, it's totally fine, if it's just for aesthetics, then I suggest using it once (and if you feel strongly about number of spaces, change the `emitIndent` function :) ).

Reply:
Changed

Reply:
removed